File: mobile/src/main/java/com/ataulm/whatsnext/film/FilmDisplayer.java
package com.ataulm.whatsnext.film;

import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.ataulm.whatsnext.Film;
import com.bumptech.glide.Glide;

class FilmDisplayer {

    private final ImageView posterImageView;
    private final TextView titleTextView;
    private final TextView watchStatusTextView;
    private final Button markAsWatchedButton;
    private final Button markAsNotWatchedButton;

    FilmDisplayer(ImageView posterImageView, TextView titleTextView, TextView watchStatusTextView, Button markAsWatchedButton, Button markAsNotWatchedButton) {
        this.posterImageView = posterImageView;
        this.titleTextView = titleTextView;
        this.watchStatusTextView = watchStatusTextView;
        this.markAsWatchedButton = markAsWatchedButton;
        this.markAsNotWatchedButton = markAsNotWatchedButton;
    }

    void attach(final Callback callback) {
        markAsWatchedButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                callback.onClickMarkAsWatched();
            }
        });

        markAsNotWatchedButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                callback.onClickMarkAsNotWatched();
            }
        });
    }

    void detachCallback() {
        markAsWatchedButton.setOnClickListener(null);
        markAsNotWatchedButton.setOnClickListener(null);
    }

    void display(Film film) {
        titleTextView.setText(film.getSummary().getName());
        watchStatusTextView.setText(film.getRelationship().getWatched() ? "watched" : "not watched");
        Glide.with(posterImageView.getContext())
                .load(film.getSummary().getPoster().forWidth(posterImageView.getWidth()).getUrl())
                .into(posterImageView);
    }

    interface Callback {

        void onClickMarkAsWatched();

        void onClickMarkAsNotWatched();
    }
}

File: mobile/src/main/java/com/ataulm/whatsnext/film/FilmDisplayer.kt
package com.ataulm.whatsnext.film

import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import com.ataulm.whatsnext.Film
import com.bumptech.glide.Glide

internal class FilmDisplayer(
        private val posterImageView: ImageView,
        private val titleTextView: TextView,
        private val watchStatusTextView: TextView,
        private val markAsWatchedButton: Button,
        private val markAsNotWatchedButton: Button
) {

    fun attach(callback: Callback) {
        markAsWatchedButton.setOnClickListener {
            callback.onClickMarkAsWatched()
        }
        markAsNotWatchedButton.setOnClickListener {
            callback.onClickMarkAsNotWatched()
        }
    }

    fun detachCallback() {
        markAsWatchedButton.setOnClickListener(null)
        markAsNotWatchedButton.setOnClickListener(null)
    }

    fun display(film: Film) {
        titleTextView.text = film.summary.name
        watchStatusTextView.text = if (film.relationship.watched) "watched" else "not watched"

        Glide.with(posterImageView.context)
                .load(film.summary.poster.bestFor(posterImageView.width, posterImageView.height)?.url)
                .into(posterImageView)
    }

    internal interface Callback {

        fun onClickMarkAsWatched()

        fun onClickMarkAsNotWatched()
    }
}
